packages/util/src/error.ts

```diff
@@ -6,3 +6,24 @@ export function errorWithCause(message: string, cause: Error): Error {
   }
   return newError;
 }
+
+export class ErrorWithCause extends Error {
+  constructor(message: string, public cause?: Error) {
+    super(message); // 'Error' breaks prototype chain here
+    Object.setPrototypeOf(this, new.target.prototype); // restore prototype chain
+    this.name = 'ErrorWithCause';
+
+    // Stack is a non-standard property according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+    if (this.stack && cause?.stack) {
+      this.stack = this.stack + '\nCaused by:\n' + cause.stack;
+    }
+  }
+
+  get rootCause(): Error {
+    return isErrorWithCause(this.cause) ? this.cause.rootCause : this.cause || this;
+  }
+}
+
+export function isErrorWithCause(err: Error | undefined): err is ErrorWithCause {
```

Comment:
```suggestion
export function isErrorWithCause(err?: Error): boolean {
```
I dont know this `:err is ErrorWithCause` syntax. What does this do?

Reply:
Nevermind, I think I figured it out. Depending on the input, the return value will be known at compile time, right?

Reply:
Yes. It is called a type guard.